File: python/solvers/cplex_lp.py


import time

from ._common import make_result, memory_monitor, read_graph

CPLEX_COMMUNITY_VAR_LIMIT = 1000


def solve(graph_path: str, time_limit: float = 300.0) -> dict:
    try:
        import cplex
    except ImportError as e:
        return make_result(status="error", error_message=f"cplex import failed: {e}")

    with memory_monitor() as peak_mb:
        t0 = time.perf_counter()
        try:
            n_left, n_right, edges = read_graph(graph_path)
            m = len(edges)

            if m > CPLEX_COMMUNITY_VAR_LIMIT:
                return make_result(
                    time_seconds=time.perf_counter() - t0,
                    peak_memory_mb=peak_mb[0],
                    status="skipped",
                    error_message=(
                        f"CPLEX Community Edition: m={m} variables exceeds "
                        f"{CPLEX_COMMUNITY_VAR_LIMIT}-variable license cap"
                    ),
                )

            cpx = cplex.Cplex()
            cpx.set_log_stream(None)
            cpx.set_error_stream(None)
            cpx.set_warning_stream(None)
            cpx.set_results_stream(None)
            cpx.parameters.timelimit.set(time_limit)
            cpx.parameters.lpmethod.set(2)        
            cpx.objective.set_sense(cpx.objective.sense.maximize)
            cpx.variables.add(obj=[1.0]*m, lb=[0.0]*m, ub=[1.0]*m)

            left_cols = [[] for _ in range(n_left)]
            right_cols = [[] for _ in range(n_right)]
            for idx, (u, w) in enumerate(edges):
                left_cols[u].append(idx)
                right_cols[w].append(idx)

            rows = []
            for u in range(n_left):
                if left_cols[u]:
                    rows.append([[left_cols[u], [1.0]*len(left_cols[u])], "L", 1.0])
            for w in range(n_right):
                if right_cols[w]:
                    rows.append([[right_cols[w], [1.0]*len(right_cols[w])], "L", 1.0])
            if rows:
                cpx.linear_constraints.add(
                    lin_expr=[r[0] for r in rows],
                    senses=[r[1] for r in rows],
                    rhs=[r[2] for r in rows],
                )

            cpx.solve()
            elapsed = time.perf_counter() - t0
            stat = cpx.solution.get_status()

            
            if stat in (11, 25, 107):
                return make_result(
                    time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                    status="timeout",
                    error_message=f"CPLEX hit time limit (status {stat})",
                )
            if stat not in (1, 5):
                return make_result(
                    time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                    status="error",
                    error_message=f"CPLEX non-optimal status {stat}",
                )

            vals = cpx.solution.get_values()
            for i, v in enumerate(vals):
                if abs(v - round(v)) > 1e-6:
                    return make_result(
                        time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                        status="error",
                        error_message=f"Non-integral LP value {v} at edge {i}",
                    )

            return make_result(
                matching_size=int(round(sum(vals))),
                time_seconds=elapsed, peak_memory_mb=peak_mb[0],
                status="optimal",
            )
        except Exception as e:
            return make_result(
                time_seconds=time.perf_counter() - t0,
                peak_memory_mb=peak_mb[0],
                status="error",
                error_message=str(e),
            )

File: src/cplex_lp.cpp

#include <chrono>
#include <cstdio>
#include <string>
#include <vector>

#include <ilcplex/cplex.h>

#include "graph_io.hpp"

static constexpr int CPLEX_COMMUNITY_VAR_LIMIT = 1000;

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: cplex_lp <graph_file> [--time-limit SECONDS]\n");
        return 1;
    }
    double time_limit = parse_time_limit(argc, argv, 300.0);

    Graph g;
    try {
        g = read_graph(argv[1]);
    } catch (const std::exception& e) {
        print_result(-1, 0.0, "error", e.what());
        return 1;
    }
    int m = (int)g.edges.size();

    if (m > CPLEX_COMMUNITY_VAR_LIMIT) {
        char buf[160];
        snprintf(buf, sizeof(buf),
                 "CPLEX Community Edition: m=%d exceeds %d-variable license cap",
                 m, CPLEX_COMMUNITY_VAR_LIMIT);
        print_result(-1, 0.0, "skipped", buf);
        return 0;
    }

    int status_code = 0;
    CPXENVptr env = CPXopenCPLEX(&status_code);
    if (!env) {
        char buf[64]; snprintf(buf, sizeof(buf), "CPXopenCPLEX failed %d", status_code);
        print_result(-1, 0.0, "error", buf);
        return 1;
    }
    CPXsetintparam(env, CPXPARAM_ScreenOutput, CPX_OFF);
    CPXsetdblparam(env, CPXPARAM_TimeLimit, time_limit);
    CPXsetintparam(env, CPXPARAM_LPMethod, CPX_ALG_DUAL);

    CPXLPptr lp = CPXcreateprob(env, &status_code, "mcbp");
    if (!lp) {
        print_result(-1, 0.0, "error", "CPXcreateprob failed");
        CPXcloseCPLEX(&env);
        return 1;
    }
    CPXchgobjsen(env, lp, CPX_MAX);

    std::vector<double> obj(m, 1.0), lb(m, 0.0), ub(m, 1.0);
    CPXnewcols(env, lp, m, obj.data(), lb.data(), ub.data(), nullptr, nullptr);

    std::vector<std::vector<int>> left_cols(g.n_left), right_cols(g.n_right);
    for (int i = 0; i < m; ++i) {
        left_cols[g.edges[i].first].push_back(i);
        right_cols[g.edges[i].second].push_back(i);
    }

    std::vector<int> rmatbeg, rmatind;
    std::vector<double> rmatval, rhs;
    std::vector<char> sense;
    int nz = 0;
    auto add_row = [&](const std::vector<int>& cols) {
        rmatbeg.push_back(nz);
        for (int c : cols) {
            rmatind.push_back(c); rmatval.push_back(1.0); ++nz;
        }
        rhs.push_back(1.0); sense.push_back('L');
    };
    for (int u = 0; u < g.n_left;  ++u) if (!left_cols[u].empty())  add_row(left_cols[u]);
    for (int w = 0; w < g.n_right; ++w) if (!right_cols[w].empty()) add_row(right_cols[w]);

    int num_rows = (int)rmatbeg.size();
    if (num_rows > 0) {
        CPXaddrows(env, lp, 0, num_rows, nz,
                   rhs.data(), sense.data(),
                   rmatbeg.data(), rmatind.data(), rmatval.data(),
                   nullptr, nullptr);
    }

    auto t0 = std::chrono::steady_clock::now();
    int solve_err = CPXlpopt(env, lp);
    auto t1 = std::chrono::steady_clock::now();
    double elapsed = std::chrono::duration<double>(t1 - t0).count();

    if (solve_err) {
        char buf[64]; snprintf(buf, sizeof(buf), "CPXlpopt error %d", solve_err);
        print_result(-1, elapsed, "error", buf);
        CPXfreeprob(env, &lp); CPXcloseCPLEX(&env);
        return 1;
    }

    int stat = CPXgetstat(env, lp);
    
    if (stat == 11 || stat == 25 || stat == 107) {
        char buf[80]; snprintf(buf, sizeof(buf), "CPLEX hit time limit (status %d)", stat);
        print_result(-1, elapsed, "timeout", buf);
        CPXfreeprob(env, &lp); CPXcloseCPLEX(&env);
        return 0;
    }
    if (stat != 1 && stat != 5) {
        char buf[80]; snprintf(buf, sizeof(buf), "CPLEX non-optimal status %d", stat);
        print_result(-1, elapsed, "error", buf);
        CPXfreeprob(env, &lp); CPXcloseCPLEX(&env);
        return 1;
    }

    double objval = 0.0;
    CPXgetobjval(env, lp, &objval);
    print_result((int)(objval + 0.5), elapsed, "optimal");
    CPXfreeprob(env, &lp);
    CPXcloseCPLEX(&env);
    return 0;
}
